jazkarta/shop/browser/checkout.py

```diff
@@ -24,13 +32,318 @@
+            contact_info['last_name'] = self.request.form['x_last_name']
+        if 'x_email' in self.request.form:
+            contact_info['email'] = self.request.form['x_email']
+        if 'x_phone' in self.request.form:
+            contact_info['phone'] = self.request.form['x_phone']
+        if 'x_address' in self.request.form:
+            contact_info['address'] = self.request.form['x_address']
+        if 'x_city' in self.request.form:
+            contact_info['city'] = self.request.form['x_city']
+        if 'x_state' in self.request.form:
+            contact_info['state'] = self.request.form['x_state']
+        if 'x_zip' in self.request.form:
+            contact_info['zip'] = self.request.form['x_zip']
+        if 'x_country' in self.request.form:
+            contact_info['country'] = self.request.form['x_country']
+
+        method = 'Online Payment'
+        charge_result = {'success': True}
+
+        if not self.error:
+            try:
+                self.store_order(method, charge_result, userid, contact_info)
+                self.clear_cart()
+            except ConflictError:
+                raise Exception(
+                    'Failed to store results of payment after multiple '
+                    'retries. Please contact us for assistance.')
+
+    # This code runs after the payment is processed
+    # to update various data in the ZODB.
+    @run_in_transaction(retries=10)
+    def store_order(self, method, charge_result, userid, contact_info):
+        """Store various data following payment processing.
+
+        This code runs in a separate transaction (with up to 10 retries)
+        to make sure that ConflictErrors do not cause the Zope publisher
+        to retry the request which performs payment processing.
+
+        # ^ is the above still true when using Authorize.Net SIM??
+
+        IMPORTANT: Make sure that code which runs before this is called
+        does not try to make persistent changes, because they'll be lost.
+        """
+        order = self.cart.data
+        order['bill_to'] = contact_info
+        order['notes'] = self.request.form.get('notes')
+
+        # `success` is only present for purchases, not refunds
+        if 'success' in charge_result and charge_result['success']:
+
+            # Call `after_purchase` hook for each product
+            coupons_used = set()
+            for index, item in enumerate(self.cart.items):
+                ob = resolve_uid(item.uid)
+                if ob is not None:
+                    purchase_handler = IPurchaseHandler(ob)
+                    purchase_handler.after_purchase(item._item)
+
+                # keep track of coupons used
+                if item.is_discounted:
+                    coupons_used.add(item.coupon)
+
+            # store count of coupon usage
+            for coupon_uid in coupons_used:
+                storage.increment_shop_data(
+                    [userid, 'coupons', coupon_uid], 1)
```

Comment:
Some of this post-checkout handling has no tie to the payment processor. Can we avoid duplicating it? Otherwise I'm worried that we'll fix a bug in one copy but forget to fix the other one.